File: C++ Project/.vscode/Recursion/mergeSort.cpp
#include <iostream>
using namespace std;

void merge2SortedArray(int arr[], int start, int end)
{

    int mid = start + (end - start) / 2;

    int length1 = mid - start + 1, length2 = end - mid;

    int *first = new int[length1];
    int *second = new int[length2];

    int mainArrayIndex = start;
    for (int i = 0; i < length1; i++)
        first[i] = arr[mainArrayIndex++];

    mainArrayIndex = mid + 1;
    for (int i = 0; i < length2; i++)
        second[i] = arr[mainArrayIndex++];

    // merge 2 sorted Array
    int index1 = 0, index2 = 0, mainArrayIndex = start;

    while (index1 < length1 && index2 < length2)
    {
        if (first[index1] <= second[index2])
            arr[mainArrayIndex++] = first[index1++];
        else
            arr[mainArrayIndex++] = second[index2++];
    }

    while (index1 < length1)
        arr[mainArrayIndex++] = first[index1++];

    while (index2 < length2)
        arr[mainArrayIndex++] = second[index2++];

    delete first;
    delete second;
}

void mergeSort(int arr[], int start, int end)
{
    // base case
    if (start >= end)
        return;

    int mid = start + (end - start) / 2;

    // sort Left Part
    mergeSort(arr, start, mid);

    // sort Right part
    mergeSort(arr, mid + 1, end);

    // merge
    merge2SortedArray(arr, start, end);
}

int main(int argc, char const *argv[])
{

    return 0;
}

File: DSA in Java and Kotlin/src/kotlinLanguage/MergeSort.kt
package kotlinLanguage

fun main() {

    val demoArray = intArrayOf(38, 27, 43, 3, 9, 82, 10)

    mergeSort(demoArray, 0, demoArray.lastIndex)

    print("\n \n Demo Array ")
    demoArray.forEach { print("$it ") }

    println()
}

fun mergeSort(nums: IntArray, s: Int, e: Int) {
    //todo:  base case
    if (s >= e) return

    val mid = s + (e - s) / 2

    // left part
    mergeSort(nums, s, mid)
    //right part
    mergeSort(nums, mid + 1, e)

    merge(nums, s, e)
}

fun merge(nums: IntArray, s: Int, e: Int) {
    val mid = s + (e - s) / 2

    //todo: break array into two part
    val arr1 = ArrayList<Int>(mid - s + 1)
    val arr2 = ArrayList<Int>(e - mid)

    //todo: create two new array separated form original from mid
    var originalArrayIndex = s
    for (i in 0..<mid - s + 1) arr1.add(nums[originalArrayIndex++])
    for (i in 0..<e - mid) arr2.add(nums[originalArrayIndex++])

    //todo: merge both array
    var index1 = 0
    var index2 = 0

    originalArrayIndex = s

    while (index1 < arr1.size && index2 < arr2.size) {
        if (arr1[index1] < arr2[index2]) {
            nums[originalArrayIndex++] = arr1[index1++]
        } else nums[originalArrayIndex++] = arr2[index2++]
    }

    while (index1 < arr1.size) nums[originalArrayIndex++] = arr1[index1++]
    while (index2 < arr2.size) nums[originalArrayIndex++] = arr2[index2++]
}

fun printArray(arr: ArrayList<Int>) {
    print("Array ")
    arr.forEach { print("$it ") }
}

fun printArray(arr: IntArray) {
    print("Array ")
    arr.forEach { print("$it ") }
}